zio-http/src/main/scala/zio/http/netty/client/NettyConnectionPool.scala

```diff
@@ -114,8 +108,107 @@ object NettyConnectionPool {
       result        <- ZIO.attempt(channelFuture.channel())
     } yield result
+
+    for {
+      resolvedHosts <- dnsResolver.resolve(location.host)
+      sortedHosts    = sortHostsHappyEyeballs(resolvedHosts)
+      createChannels = sortedHosts.map(createChannel).toList
+      // Racing them all works but is not HappyEyeballs.
+      // result <- ZIO.raceAll(ZIO.never, createChannels)
+      result <- happyEyeballsWithRelease(
+        createChannels,
+        250.millis,
+        (jchannel: JChannel) => NettyFutureExecutor.executed(jchannel.closeFuture()).ignore,
+      )
+    } yield result
+  }
+
+  /**
+   * Sort addresses according to Happy Eyeballs Version 2 RFC.
+   *
+   * See [[https://www.rfc-editor.org/rfc/rfc8305#section-4]]
+   */
+  private final def sortHostsHappyEyeballs(hosts: Chunk[InetAddress]): Chunk[InetAddress] = {
+    val (ipv6, ipv4) = hosts.partition {
+      case _: Inet6Address => true
+      case _               => false
+    }
+
+    if (ipv6.isEmpty || ipv4.isEmpty) {
+      hosts
+    } else {
+      // Interleave the two address families.
+      val ipv6Iterator = ipv6.chunkIterator
+      val ipv4Iterator = ipv4.chunkIterator
+
+      val builder = ChunkBuilder.make[InetAddress]()
+      builder.sizeHint(ipv4.size + ipv6.size)
+
+      var ipv6Index   = 0
+      var ipv4Index   = 0
+      var ipv6HasNext = false
+      var ipv4HasNext = false
+      while ({
+        ipv6HasNext = ipv6Iterator.hasNextAt(ipv6Index)
+        ipv4HasNext = ipv4Iterator.hasNextAt(ipv4Index)
+        ipv6HasNext || ipv4HasNext
+      }) {
+        if (ipv6HasNext) {
+          builder += ipv6Iterator.nextAt(ipv6Index)
+          ipv6Index += 1
+        }
+        if (ipv4HasNext) {
+          builder += ipv4Iterator.nextAt(ipv4Index)
+          ipv4Index += 1
+        }
+      }
+      builder.result()
+    }
   }
 
+  /**
+   * Attempt to execute the tasks according to Happy Eyeballs Version 2 RFC.
+   *
+   * Tasks are not started simultaneously, rather they are started in sequence.
+   *
+   * The next task is started if the current task fails or after the specified
+   * delay, whichever comes first.
+   *
+   * See [[https://www.rfc-editor.org/rfc/rfc8305#section-5]]
+   *
+   * Returns the first successful connection. Fails with an exception if none
+   * succeed.
+   */
+  private def executeHappyEyeballs[A](tasks: List[Task[A]], delay: Duration): Task[A] =
+    tasks match {
+      // What should this error be?
+      case Nil                => ZIO.fail(new IllegalStateException("No tasks left."))
+      case task :: Nil        => task
+      case task :: otherTasks =>
+        Promise.make[Nothing, Unit].flatMap { failurePromise =>
+          val taskWithFailure = task.onError(_ => failurePromise.complete(ZIO.unit).ignore)
+          val continue        = failurePromise.await.timeout(delay).ignore *> executeHappyEyeballs(otherTasks, delay)
+          taskWithFailure.race(continue)
```

Comment:
In the NettyConnectionPoolSpec tests `taskWithFailure` is completing, but the function does not terminate. I am not sure why the race on the right `continue` is not being interrupted. I am sure I'm doing something wrong but I don't understand what exactly